File: tests-python/pages/home_page.py
from __future__ import annotations

import allure
from selenium.webdriver.common.by import By

from pages.base import BasePage


class HomePage(BasePage):
    LAYOUT = (By.CSS_SELECTOR, "[data-testid='reference-layout']")
    HEALTH = (By.CSS_SELECTOR, "[data-testid='health-status']")
    ITEMS = (By.CSS_SELECTOR, "[data-testid='items-list']")
    WELCOME = (By.CSS_SELECTOR, "[data-testid='welcome-message']")
    WELCOME_PANEL = (By.CSS_SELECTOR, "[data-testid='welcome-panel']")
    LOGOUT = (By.CSS_SELECTOR, "[data-testid='logout-button']")

    @allure.step("Open home page")
    def open_page(self) -> HomePage:
        self.open_path("/")
        return self

    @allure.step("Verify home layout is mounted")
    def should_show_layout(self) -> HomePage:
        self.wait_visible(*self.LAYOUT)
        self.wait_visible(*self.ITEMS)
        return self

    @allure.step("Verify health status contains: {text_fragment}")
    def should_show_health_text(self, text_fragment: str) -> HomePage:
        self.wait_text_contains(*self.HEALTH, text_fragment)
        return self

    @allure.step("Verify items list contains: {text_fragment}")
    def should_show_item_text(self, text_fragment: str) -> HomePage:
        self.wait_text_contains(*self.ITEMS, text_fragment)
        return self

    @allure.step("Verify welcome message: {message}")
    def should_have_welcome_message(self, message: str) -> HomePage:
        self.wait_visible(*self.WELCOME_PANEL)
        self.wait_text_contains(*self.WELCOME, message)
        return self

    @allure.step("Click logout button")
    def click_logout_button(self):
        from pages.login_page import LoginPage
        from urllib.parse import urlparse

        self.js_click(*self.LOGOUT)

        def _on_login(driver):
            path = urlparse(driver.current_url).path
            return path.rstrip("/").endswith("login")

        self.wait().until(_on_login)
        return LoginPage(self.driver, self.config)

File: tests-selenium/src/test/java/pages/HomePage.java
package pages;

import config.TestConfig;
import io.qameta.allure.Step;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class HomePage extends BasePage {

    private static final By LAYOUT = By.cssSelector("[data-testid='reference-layout']");
    private static final By HEALTH = By.cssSelector("[data-testid='health-status']");
    private static final By ITEMS = By.cssSelector("[data-testid='items-list']");
    private static final By WELCOME = By.cssSelector("[data-testid='welcome-message']");
    private static final By WELCOME_PANEL = By.cssSelector("[data-testid='welcome-panel']");
    private static final By LOGOUT = By.cssSelector("[data-testid='logout-button']");

    public HomePage(WebDriver driver, TestConfig config) {
        super(driver, config);
    }

    @Step("Open home page")
    public HomePage openPage() {
        openPath("/");
        return this;
    }

    @Step("Verify home layout is mounted")
    public HomePage shouldShowLayout() {
        waitVisible(LAYOUT);
        waitVisible(ITEMS);
        return this;
    }

    @Step("Verify health status contains: {textFragment}")
    public HomePage shouldShowHealthText(String textFragment) {
        waitTextContains(HEALTH, textFragment);
        return this;
    }

    @Step("Verify items list contains: {textFragment}")
    public HomePage shouldShowItemText(String textFragment) {
        waitTextContains(ITEMS, textFragment);
        return this;
    }

    @Step("Verify welcome message: {message}")
    public HomePage shouldHaveWelcomeMessage(String message) {
        waitVisible(WELCOME_PANEL);
        waitTextContains(WELCOME, message);
        return this;
    }

    @Step("Click logout button")
    public LoginPage clickLogoutButton() {
        jsClick(LOGOUT);
        waitUrlPathEndsWith("login");
        return new LoginPage(driver, config);
    }
}
